src/biom_subsampled.cpp

```diff
@@ -106,22 +111,162 @@ void linked_sparse_transposed::transposed_subsample_with_replacement(const uint3
+    : data_in(_data_in)
+    , idx(_idx)
+    , cnt(_cnt)
+    {}
+
+    reference operator*() const { return idx; }
+    pointer operator->() const { return &idx; }
+
+    WeightedSampleIterator& operator++()
+    {  
+       cnt++;
+       if (cnt>=data_in[idx]) {
+         cnt = 0;
+         idx++;
+       }
+       return *this;
+    }
+
+    WeightedSampleIterator operator++(int) { WeightedSampleIterator tmp = *this; ++(*this); return tmp; }
+
+    friend bool operator== (const WeightedSampleIterator& a, const WeightedSampleIterator& b)
+    {
+       return (a.data_in == b.data_in) && (a.idx == b.idx) && (a.cnt==b.cnt);
+    };
+
+    friend bool operator!= (const WeightedSampleIterator& a, const WeightedSampleIterator& b)
+    {
+       return !((a.data_in == b.data_in) && (a.idx == b.idx) && (a.cnt==b.cnt));
+    };
+
+    friend int64_t operator-(const WeightedSampleIterator& b, const WeightedSampleIterator& a)
+    {
+       int64_t diff = 0;
+       //assert(a.data_in == b.data_in);
+       //assert(a.idx <= b.idx);
+       //assert((a.idx > b.idx) || (a.cnt<=b.cnt));
+
+       for (uint32_t i = a.idx; i<b.idx; i++) {
+          diff += a.data_in[i];
+       }
+
+       return diff + b.cnt - a.cnt;
+    };
+
+private:
+
+    uint64_t *data_in;
+    uint32_t idx; // index of data_in
+    uint64_t cnt; // how deep in data_in[idx] are we (must be < data_in[idx])
+};
+
+class WeightedSample
+{
+public:
+    WeightedSample(uint32_t _max_count, uint32_t _n, uint32_t random_seed)
+    : max_count(_max_count)
+    , n(_n)
+    , generator(random_seed)
+    , current_count(0)
+    , data(max_count)
+    , sample_out(n)
+    , data_out(max_count)
+    {}
+
+   void do_sample(unsigned int length, double* *data_arr) {
+        for (unsigned int j=0; j<length; j++) data_out[j] = 0;
+
+        // note: We are assuming length>=n
+        //      Enforced by the caller (via filtering)
+        assign(length,data_arr);
+        std::sample(begin(), end(),
+                    sample_out.begin(), n,
+                    generator);
+
+        for (uint32_t j=0; j<n; j++) data_out[sample_out[j]]++;
+
+        for (unsigned int j=0; j<length; j++) *(data_arr[j]) = data_out[j];
+    }
+
+private:
+    void assign(uint32_t length, double * *data_arr) {
+       current_count = length;
+       for (uint32_t j=0; j<length; j++) data[j] = *(data_arr[j]);
+    }
+
+    WeightedSampleIterator begin() { return WeightedSampleIterator(data.data(),0,0); }
+    WeightedSampleIterator end()   { return WeightedSampleIterator(data.data(),current_count,0); } // current_count is out of bounds
+
+    uint32_t max_count;
+    uint32_t n;
+    std::mt19937 generator;
+
+    uint32_t current_count;
+
+    // use persistent buffer to minimize allocation costs
+    std::vector<uint64_t> data;  // original values
+    std::vector<uint32_t> sample_out;     // random output buffer
+    std::vector<uint32_t> data_out; // computed values
```

Comment:
Does this mean the max count for a given feature in a given sample is 2**32?

Reply:
Yes, because n in 32-bit, so the output cannot be larger than n.